### ISSUE TITLE
Async keywords are not stopped when execution is stopped gracefully

### ISSUE THREAD

**Issue by jaltendorfer** [robotframework/robotframework#4808]
Stop a test execution during an async keyword with SIGINT or CTRL+C. Now the teardown with async keywords is executed, but the test keyword still runs and is not stopped.

**Example:**

Test python library:
```
import asyncio
from robot.api.deco import library, keyword
from robot.api import logger


@library
class TestLibrary:

    @keyword(name="Async Sleep")
    async def async_sleep(self, time: int, log_message: str):
        logger.info(f"{log_message} start", also_console=True)
        await asyncio.sleep(time)
        logger.info(f"{log_message} end", also_console=True)
```
Test robot suite:
```
*** Settings ***
Library    TestLibrary.py


*** Test Cases ***
Test
    TestLibrary.Async Sleep    ${5}    Sleep Test
    [Teardown]    TestLibrary.Async Sleep    ${10}    Sleep Teardown
```

Result:

![image](https://github.com/robotframework/robotframework/assets/138000046/848f7cb1-e359-44ba-b8f0-d77788e7544a)

**Comment by pekkaklarck** [comment]
Teardowns are, by default, executed when execution is gracefully stopped. If you don't want that, you can use the `--skip-teardown-on-exit` command line option. Does that work for you? Do you see differences with normal and async keywords in this regard?

**Comment by jaltendorfer** [comment]
The problem is not that the teardown is executed. The problem is, that the async keyword in the test case is not stopped immediately before executing the teardown. In the example result of the post before, you can see the log messages of the test case keyword in the teardown after a graceful stop of the execution.

Normal behaviour with graceful stop:
- Running keyword in the test case is stopped immediately
- Teardowns are executed

Current behaviour when using async keywords:
- Running keyword in the test case is not stopped
- Teardowns are executed and the test case keyword is still running (in background) until completion.

In my example in the post above, you can see the log message "Sleep Test" being logged in the teardown, although the execution was stopped and the testcase keyword should have been stopped immediately (only the teardown should be executed).

**Comment by pekkaklarck** [comment]
I see. Apparently async methods would need separate handling to get them killed. Would be nice to get that fixed in RF 6.1.2.

I'm on holiday and won't have time to look at this in the near future. If someone with more async knowledge would be interested to look at this already now that would be great. Ping @ygorpontelo.

**Comment by pekkaklarck** [comment]
Sorry for forgetting about this. The PR looks good otherwise but we'd also need a test. I hope we could still get this into RF 7.

**Comment by pekkaklarck** [comment]
After thinking this a bit more, the current behavior is buggy and we believe bugs should in general be fixed as soon as possible. I added this to RF 7 scope. Hopefully you @ygorpontelo can help with tests. Getting somewhat related #4803 into RF 7 would be good too.

**Comment by pekkaklarck** [comment]
Fixed by PR #4835. Great work @ygorpontelo!
